Update license issuing to accept a site_id

oikeme opened the pull request:
Update issuing a license to allow for a site_id to be passed. This will cause the issued license to be automatically assigned to the site_id that is passed and bypass the step to assign the license.

#### Changes proposed in this Pull Request

* An option `site_id` query param can now be passed to the `issue-license` endpoint. If the `site_id` is present the endpoint will assign the license to the site immediately and bypass the assignment step. A redirect to the existing final step that displays the license will be done after the license is assigned

#### Testing instructions

<!--
Add as many details as possible to help others reproduce the issue and test the fix.
"Before / After" screenshots can also be very helpful when the change is visual.
-->

* Apply this patch and start calypso first `yarn start` followed by Jetpack cloud `yarn start-jetpack-cloud-p`
* Visit the licensing dashboard and click the button to Issue A New License.
* Add the query param `site_id={your_site_id}` to the query string. The `site_id` is the blog id for the site that the license will be assigned to.
* Select the product you would like to issue the license for and issue the license. 
* Ensure that you are redirected to `licenses` page and the new license is displayed and not the `assign-licenses` page to select your site.
* Repeat the steps above without the optional param and ensure that you are able to get to the `assign-license` page, select a site and have the license assigned to the site.

<!--
Link a related issue to this PR. If the PR does not immediately resolve the issue,
for example, it requires a separate deployment to production, avoid
using the "fixes" keyword and instead attach the [Status] Fix Inbound label to
the linked issue.
-->

Related to #

oikeme committed:
Update license issuing to accept a site_id and automatically assign the license to the site_id

Update issuing a license to allow for a site_id to be passed. This will cause the issued license to be automatically assigned to the site_id that is passed and bypass the step to assign the license.

rcanepa commented on client/jetpack-cloud/sections/partner-portal/primary/issue-license/index.tsx:
```diff
@@ -7,7 +7,11 @@ import IssueLicenseForm from 'calypso/jetpack-cloud/sections/partner-portal/issu
 import AssignLicenseStepProgress from '../../assign-license-step-progress';
 
-export default function IssueLicense(): ReactElement {
+interface Props {
+	site_id?: number;
+}
+
+export default function IssueLicense( { site_id }: Props ): ReactElement {
```
Can we rename this prop to `siteId`? It is fine to use `site_id` for a query parameter, but it is better to use camel case when we pass it down to a component.

rcanepa commented on client/jetpack-cloud/sections/partner-portal/controller.tsx:
```diff
@@ -83,9 +83,10 @@ export function licensesContext( context: PageJS.Context, next: () => void ): vo
 }
 
 export function issueLicenseContext( context: PageJS.Context, next: () => void ): void {
+	const { site_id } = context.query;
```
Can we rename the `site_id` variable to `siteId`? In practice, this would look like this:
```
const { site_id: siteId } = context.query; // we rename site_id to siteId
```

rcanepa replied:
I forgot to mention that on Calypso, the convention is to use camel case as much as possible.

oikeme replied:
I've updated this and other instances from site_id to siteId

rcanepa commented on client/jetpack-cloud/sections/partner-portal/controller.tsx:
```diff
@@ -83,9 +83,10 @@ export function licensesContext( context: PageJS.Context, next: () => void ): vo
 	context.header = <Header />;
 	context.secondary = <PartnerPortalSidebar path={ context.path } />;
-	context.primary = <IssueLicense />;
+	context.primary = <IssueLicense site_id={ site_id } />;
```
Here, we would have to change `<IssueLicense site_id={ site_id } />` to `<IssueLicense siteId={ siteId } />`?

rcanepa commented:
These changes work as expected. Should we restrict the license assignment process only to siteIds we know are linked to the current logged-in WordPress.com user? What could happen if I pass a site Id that does not belong to me?

oikeme committed:
Update variable name

Update query param and uses from site_id to siteId

oikeme committed:
Check site id ownership when issuing a license

Update the controller to verify that the passed site id is one of the users sites before using it when issuing a license.

oikeme commented:
> These changes work as expected. Should we restrict the license assignment process only to siteIds we know are linked to the current logged-in WordPress.com user? What could happen if I pass a site Id that does not belong to me? 

@rcanepa good catch. I believe this would cause an error with the API but it definitely makes more sense to check before we attempt the request. I've updated things to make sure the site id is in the users sites before using it.

oikeme committed:
Rename query param

Rename the site_id query param

rcanepa commented:
> > These changes work as expected. Should we restrict the license assignment process only to siteIds we know are linked to the current logged-in WordPress.com user? What could happen if I pass a site Id that does not belong to me?
> 
> @rcanepa good catch. I believe this would cause an error with the API but it definitely makes more sense to check before we attempt the request. I've updated things to make sure the site id is in the users sites before using it.

Maybe, we could disregard the `siteId` and take the user to the next step where they have to select a site (the current flow).

oikeme commented:
> > > These changes work as expected. Should we restrict the license assignment process only to siteIds we know are linked to the current logged-in WordPress.com user? What could happen if I pass a site Id that does not belong to me?
> > 
> > 
> > @rcanepa good catch. I believe this would cause an error with the API but it definitely makes more sense to check before we attempt the request. I've updated things to make sure the site id is in the users sites before using it.
> 
> Maybe, we could disregard the `siteId` and take the user to the next step where they have to select a site (the current flow).

@rcanepa apologies I didn't make that clear. If the siteId is not in the users sites then it's set to null which will allow the user to continue on to the current flow where they can select one of their sites.

rcanepa commented on client/jetpack-cloud/sections/partner-portal/controller.tsx:
```diff
@@ -83,9 +84,14 @@ export function licensesContext( context: PageJS.Context, next: () => void ): vo
+	const { site_id: siteId } = context.query;
+	const state = context.store.getState();
+	const sites = getSitesItems( state );
+	const selectedSite = sites.hasOwnProperty( siteId ) ? siteId : null;
```
I recommend replacing `sites.hasOwnProperty( siteId )` with `sites[ siteId ]` for simplicity's sake.

Also, returning `null` will not conform to the interface of the `IssueLicense` component. The interface looks like:

```
interface Props {
	selectedSite?: number;
}
```

In this case, `selectedSite` is expected to be `undefined` or a number. If we wanted to include `null` as a possible value, we would have to change the type to `selectedSite?: number | null` (the accepted types would be `undefined`, `null`, and any number).

rcanepa reviewed (approved):
LGTM.

I left a small comment about a type mismatch. Everything looks and works well.

oikeme committed:
Update siteId allowed types

Allow null as a possible type for siteId property when attempting to automatically assign an issued license.
